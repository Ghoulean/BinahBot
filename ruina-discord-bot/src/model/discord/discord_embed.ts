/**
 * Object holding embed data.
 * 
 * See also: https://discord.com/developers/docs/resources/channel#embed-object
 */
export type DiscordEmbed = {
    title?: string;
    description?: string;
    color?: number;
    image?: {
        url: string;
    },
    footer?: {
        text: string;
        icon_url?: string;
    },
    author?: {
        name: string;
        url?: string;
        icon_url?: string;
    },
    fields: DiscordEmbedFields[];
}

export type DiscordEmbedFields = {
    name: string;
    value: string;
    inline?: boolean;
}
