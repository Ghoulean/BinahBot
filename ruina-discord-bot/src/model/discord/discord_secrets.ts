/**
 * Object holding Discord secrets stored in AWS Secrets Manager
 * 
 */
export type DiscordSecrets = {
    applicationId: string;
    authToken: string;
    publicKey: string;
}
