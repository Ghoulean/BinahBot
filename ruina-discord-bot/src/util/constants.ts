export enum DiscordInteractionTypes {
    PING = 1,
    APPLICATION_COMMAND = 2,
    MESSAGE_COMPONENT = 3,
    APPLICATION_COMMAND_AUTOCOMPLETE = 4,
    MODAL_SUBMIT = 5
}

export enum DiscordEmbedColors {
    AWAKENING_ABNO_PAGE = 0x40ce78,
    BREAKDOWN_ABNO_PAGE = 0xd14141
}
