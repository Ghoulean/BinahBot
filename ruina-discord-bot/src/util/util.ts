import { Localization } from "@ghoulean/ruina-common";

const DISCORD_LOCALE_TO_LOCALIZATION_MAP: { [key: string]: Localization } = {
    ko: Localization.KOREAN,
    ja: Localization.JAPANESE,
    "zh-CN": Localization.CHINESE_SIMPLIFIED,
    "zh-TW": Localization.CHINESE_TRADITIONAL,
    "en-US": Localization.ENGLISH,
    "en-GB": Localization.ENGLISH,
};

export abstract class Util {
    public static deserializeDiscordLocale(
        discordLocale: string
    ): Localization {
        return (
            DISCORD_LOCALE_TO_LOCALIZATION_MAP[discordLocale] ??
            Localization.ENGLISH
        );
    }
}
