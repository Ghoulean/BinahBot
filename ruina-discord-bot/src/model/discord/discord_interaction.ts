import { DiscordEmbed } from "./discord_embed";

/**
 * Object holding interaction data. This only reflects the information that the bot cares about; other fields are discarded.
 *
 * See also: https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object
 */
export type DiscordInteraction = {
    id: string;
    type: number; // TODO: enum?
    token: string;
    application_id: string;
    locale?: string;
    guild_locale?: string;
    channel_id?: string;
    data?: {
        id: string;
        name: string;
        options?: DiscordInteractionOptions[];
    };
    metadata: DiscordInteractionMetadata
};

export type DiscordInteractionOptions = {
    name: string;
    value: string;
};

export type DiscordInteractionMetadata = {
    timestamp: string;
    signature: string;
    jsonBody: string;
}

export type DiscordInteractionResponse = {
    type: number; // TODO: enum?
    data?: DiscordInteractionResponseMessage | DiscordInteractionResponseAutocomplete;
};

export type DiscordInteractionResponseMessage = {
    embeds?: DiscordEmbed[];
};

export type DiscordInteractionResponseAutocomplete = {
    choices?: DiscordInteractionOptions[];
};
