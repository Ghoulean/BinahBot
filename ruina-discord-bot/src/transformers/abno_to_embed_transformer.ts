import { DecoratedAbnoPage, Localization } from "@ghoulean/ruina-common";
import { DiscordEmbedColors } from "../util/constants";
import { DiscordEmbed } from "../model/discord/discord_embed";

export class AbnoToEmbedTransformer {
    private readonly s3BucketName: string;

    constructor(s3BucketName: string) {
        this.s3BucketName = s3BucketName;
    }

    // TODO: strongly type embeds
    // TODO: localize via abno.locale
    public transform(abno: DecoratedAbnoPage): DiscordEmbed {
        const embedColor: number = abno.emotionSign > 0 ? DiscordEmbedColors.AWAKENING_ABNO_PAGE : DiscordEmbedColors.BREAKDOWN_ABNO_PAGE;
        const abnoType: string = abno.emotionSign > 0 ? "Awakening" : "Breakdown";
        return {
            title: abno.name,
            color: embedColor,
            image: {
                // TODO: less scuffed way to construct URL.
                // TODO: imagePath is currently prefixed with `/`. Revisit
                url: `https://${this.s3BucketName}.s3.amazonaws.com${abno.imagePath}`,
            },
            fields: [
                {
                    name: "Flavor text",
                    value: abno.flavorText,
                    inline: true,
                },
                {
                    name: "Effect",
                    value: abno.description,
                    inline: true,
                },
                {
                    name: "Bias",
                    value: String(abno.emotionRate),
                    inline: true,
                },
                {
                    name: "Type",
                    value: abnoType,
                    inline: true,
                },
                {
                    name: "Emotion level",
                    value: String(abno.emoLevel),
                    inline: true,
                },
                {
                    name: "Floor",
                    value: abno.floor,
                    inline: true,
                },
            ],
        };
    }
}
