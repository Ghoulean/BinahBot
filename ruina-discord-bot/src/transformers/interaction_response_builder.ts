import { DiscordEmbed } from "../model/discord/discord_embed";
import {
    DiscordInteractionResponse,
    DiscordInteractionResponseAutocomplete,
    DiscordInteractionResponseMessage
} from "../model/discord/discord_interaction";
import { DiscordInteractionTypes } from "../util/constants";

const PING_DISCORD_INTERACTION_TYPE = 1;
const MESSAGE_DISCORD_INTERACTION_TYPE = 4;
const AUTOCOMPLETE_RESULT_DISCORD_INTERACTION_TYPE = 8;

const PING_RESPONSE: DiscordInteractionResponse = {
    type: PING_DISCORD_INTERACTION_TYPE,
};

export class InteractionResponseBuilder {
    public buildResponse(
        requestType: number,
        data: unknown
    ): DiscordInteractionResponse {
        switch (requestType) {
            case DiscordInteractionTypes.PING:
                return PING_RESPONSE;
            case DiscordInteractionTypes.APPLICATION_COMMAND:
                return this.buildCommandResponse(data as DiscordEmbed);
            case DiscordInteractionTypes.APPLICATION_COMMAND_AUTOCOMPLETE:
                return this.buildAutocompleteResponse(data as string[]);
            default:
                throw new Error(
                    `Unrecognized or unsupported request type ${requestType}`
                );
        }
    }

    private buildCommandResponse(
        embed: DiscordEmbed
    ): DiscordInteractionResponse {
        const discordPayload: DiscordInteractionResponseMessage = {
            embeds: [embed],
        };
        return {
            type: MESSAGE_DISCORD_INTERACTION_TYPE,
            data: discordPayload,
        };
    }

    private buildAutocompleteResponse(
        choices: string[]
    ): DiscordInteractionResponse {
        const discordPayload: DiscordInteractionResponseAutocomplete = {
            choices: choices.map((s: string) => {
                return {
                    name: s,
                    value: s,
                };
            }),
        };
        return {
            type: AUTOCOMPLETE_RESULT_DISCORD_INTERACTION_TYPE,
            data: discordPayload,
        };
    }
}
