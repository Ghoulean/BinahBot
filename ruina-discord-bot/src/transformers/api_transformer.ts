import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { DiscordInteraction } from "../model/discord/discord_interaction";

const TIMESTAMP_HEADER = "x-signature-timestamp";
const SIGNATURE_HEADER = "x-signature-ed25519";

export class ApiTransformer {
    constructor() {}

    public transformApiRequestToInteractionPayload(
        event: APIGatewayProxyEvent
    ): DiscordInteraction {
        return {
            ...JSON.parse(event.body!) as DiscordInteraction,
            metadata: {
                timestamp: event.headers[TIMESTAMP_HEADER]!,
                signature: event.headers[SIGNATURE_HEADER]!,
                jsonBody: event.body!,
            },
        };
    }

    public transformDataBlobToApiResponse(blob: any): APIGatewayProxyResult {
        return {
            statusCode: 200,
            headers: {
                "Content-Type": "application/json",
            },
            body: JSON.stringify(blob),
        };
    }
}
