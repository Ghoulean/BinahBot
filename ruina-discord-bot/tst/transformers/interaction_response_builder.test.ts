import { DiscordEmbed } from "../../src/model/discord/discord_embed";
import { DiscordInteractionOptions } from "../../src/model/discord/discord_interaction";
import { InteractionResponseBuilder } from "../../src/transformers/interaction_response_builder";
import { DiscordInteractionTypes } from "../../src/util/constants";

const DISCORD_EMBED: DiscordEmbed = {
    fields: [],
};

const DISCORD_AUTOCOMPLETE_LENGTH: number = 3;

const DISCORD_AUTOCOMPLETE: DiscordInteractionOptions[] = Array(
    DISCORD_AUTOCOMPLETE_LENGTH
).fill({ name: "Name", value: "value" });

const BAD_REQUEST_TYPE: number = 9999;

let interactionResponseBuilder: InteractionResponseBuilder;

beforeEach(() => {
    interactionResponseBuilder = new InteractionResponseBuilder();
});

test("should built response for pings", () => {
    expect(
        interactionResponseBuilder.buildResponse(
            DiscordInteractionTypes.PING,
            null
        )
    ).toMatchSnapshot();
});

test("should built response for commands", () => {
    expect(
        interactionResponseBuilder.buildResponse(
            DiscordInteractionTypes.APPLICATION_COMMAND,
            DISCORD_EMBED
        )
    ).toMatchSnapshot();
});

test("should built response for autocomplete", () => {
    expect(
        interactionResponseBuilder.buildResponse(
            DiscordInteractionTypes.APPLICATION_COMMAND_AUTOCOMPLETE,
            DISCORD_AUTOCOMPLETE
        )
    ).toMatchSnapshot();
});

test("should error on bad request type", () => {
    expect(() => {
        interactionResponseBuilder.buildResponse(BAD_REQUEST_TYPE, null);
    }).toThrow();
});
