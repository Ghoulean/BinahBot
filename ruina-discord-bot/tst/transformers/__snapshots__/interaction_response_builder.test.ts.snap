// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should built response for autocomplete 1`] = `
{
  "data": {
    "choices": [
      {
        "name": {
          "name": "Name",
          "value": "value",
        },
        "value": {
          "name": "Name",
          "value": "value",
        },
      },
      {
        "name": {
          "name": "Name",
          "value": "value",
        },
        "value": {
          "name": "Name",
          "value": "value",
        },
      },
      {
        "name": {
          "name": "Name",
          "value": "value",
        },
        "value": {
          "name": "Name",
          "value": "value",
        },
      },
    ],
  },
  "type": 8,
}
`;

exports[`should built response for commands 1`] = `
{
  "data": {
    "embeds": [
      {
        "fields": [],
      },
    ],
  },
  "type": 4,
}
`;

exports[`should built response for pings 1`] = `
{
  "type": 1,
}
`;
